### PR TITLE
Travis cache solr/fedora downloads

### PR DESCRIPTION
**Pull request by darrenleeweber**
Fixes slack requests for travis to cache the solr/fedora downloads, to avoid download throttling.

Changes proposed in this pull request:
* add .solr_wrapper
  * specify a download directory
* add .fcrepo_wrapper
  * specify a download directory
  * specify a version of Fedora to download

@projecthydra-labs/hyrax-code-reviewers

### CODE REVIEW

File: .solr_wrapper [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Place any default configuration for solr_wrapper here
+# port: 8983
+
+instance_dir: tmp/solr-development
+
+download_dir: tmp/solr_wrapper
+
+version: 6.5.1
```
Comment (jcoyne) [inline .solr_wrapper]: Version 1.1.0 of solr wrapper can take `latest` as a version. This is the default, so it doesn't make sense to change it here.

**Comment by cbeer** [comment]
I thought we'd already proven that having travis cache the fcrepo and solr artifacts led to a minor or negligible performance increase (just as their caching documentation suggests: https://docs.travis-ci.com/user/caching/#Things-not-to-cache)?

**Comment by darrenleeweber** [comment]
That's correct.  The purpose of caching is to lessen the load on the download providers, apparently they are throttling downloads.

File: .solr_wrapper (hunk 1 above)
Comment (darrenleeweber) [inline .solr_wrapper]: I'll push up this change once I poke around in travis to see if the cache details are working or not for the current push/pr.

**Comment by darrenleeweber** [comment]
Partial success - the first push build on travis seems to be using the new config files, i.e.
```
Loading configuration from /home/travis/build/projecthydra-labs/hyrax/.solr_wrapper
solr-6.5.1.zip: |=======================================| 100% (Time: 00:00:52 )
solr-6.5.1.zip.md5: |===================================| 100% (Time: 00:00:00 )

Loading configuration from /home/travis/build/projecthydra-labs/hyrax/.fcrepo_wrapper
fcrepo-webapp-4.7.3-RC-1-jetty-console.jar: |===========| 100% (Time: 00:00:00 )
fcrepo-webapp-4.7.3-RC-1-jetty-console.jar.md5: |=======| 100% (Time: 00:00:00 )
```

**Comment by darrenleeweber** [comment]
The cache config is also working, i.e. the first push build saves the downloads:
```
change detected (content changed, file is created, or file is deleted):

/home/travis/build/projecthydra-labs/hyrax/tmp/fcrepo_wrapper/fcrepo-webapp-4.7.3-RC-1-jetty-console.jar
/home/travis/build/projecthydra-labs/hyrax/tmp/fcrepo_wrapper/fcrepo-webapp-4.7.3-RC-1-jetty-console.jar.md5
/home/travis/build/projecthydra-labs/hyrax/tmp/solr_wrapper/solr-6.5.1.zip
/home/travis/build/projecthydra-labs/hyrax/tmp/solr_wrapper/solr-6.5.1.zip.md5
... snipped ...
```

**Comment by darrenleeweber** [comment]
When I removed the solr version from the config file, it fell back to 6.5.0 (instead of 6.5.1) and the travis build cannot download the 6.5.0 - the error in build https://travis-ci.org/projecthydra-labs/hyrax/jobs/236163848 is:
```
Loading configuration from /home/travis/build/projecthydra-labs/hyrax/.solr_wrapper

rake aborted!

SolrWrapper::SolrWrapperError: Unable to download solr from https://archive.apache.org/dist/lucene/solr/6.5.0/solr-6.5.0.zip

503 Service Unavailable: <!DOCTYPE HTML>
<html>
<head>
<meta charset="utf8">
<title>Bandwidth limit exceeded</title>
</head>
<body style="background-image:url('/gnomes.png'); background-position: top center; background-repeat: no-repeat;">
    <h1 style="text-align: center; height: 60px;">Bandwidth limit exceeded</h1>
    <p style="text-align: center;">
    The daily allowance of 5GB for this IP has been exceeded, and downloads disabled until midnight, UTC (circa 21 hours from now).<br/>
    If you have any questions about this, feel free to reach out to us at infrastructure@apache.org.
    </p>
</body>
</html>
```

That's why we need this PR to cache the downloads.

**Comment by jcoyne** [comment]
> That's why we need this PR to cache the downloads.

But if we upgrade to the latest solr wrapper, which always gets the latest version of solr, we never hit the archives' download limit.

**Comment by darrenleeweber** [comment]
Obviously travis.ci gem cache is still using an older solr_wrapper.  In any case, caching these downloads in travis will avoid download overload on the solr/fedora sources; so, why not?

**Comment by jcoyne** [comment]
I'd say don't do it because Travis-CI asks us not to cache big downloads and that we haven't seen any benefit to outweigh the additional complexity. Most certainly, apache.org is already serving solr downloads from a cache, so are we so certain that our cache is better than their cache?

**Comment by darrenleeweber** [comment]
This is marked WIP because I need to clarify how much detail should go into the `.solr_wrapper` and the `.fcrepo_wrapper` files; see TODO question in the PR description.

**Comment by darrenleeweber** [comment]
I say, let travis.ci deal with it.  They should use a decent local proxy for this kinda stuff anyway, but who cares, that's why we use travis.  Who's to say what's "large" anyway?

**Comment by darrenleeweber** [comment]
- Solr download is about 150Mb
- Fedora download is about 50Mb
- the gem cache is about 290Mb

So, what's "large"?  The downloads are about the same size as the cache for gems.  So, if the gem cache size is OK, then these downloads are not "too large".

**Comment by darrenleeweber** [comment]
I do have a concern about removing old downloads from the cache when the version changes.  I would not want to have old versions hanging around in the cache.  There is currently not automatic removal of old versions.  To manage that, we need to use a `version` value in the configs and add a rake task to the `ci` build that removes old versions.

**Comment by darrenleeweber** [comment]
This PR is good enough, as the specs were passing on Solr 6.5.1

Possible TODO later...

If the wrappers are using the hyrax configs instead of the engine_cart configs, do all the values from the engine_cart configs need to be added to the hyrax configs?  I started this PR assuming the wrapper might do a merge of all the configs, but I don't know that they are merged.  So, for example, the hyrax-solr_config has no collection details.  The engine_cart configs looks like:

```
$ cat .internal_test_app/.solr_wrapper
# Place any default configuration for solr_wrapper here
# version: 6.0.0
# port: 8983
instance_dir: tmp/solr-development
collection:
  persist: true
  dir: solr/config/
  name: hydra-development

$ cat .internal_test_app/.fcrepo_wrapper 
# Place any default configuration for solr_wrapper here
port: 8984
enable_jms: false
fcrepo_home_dir: tmp/fcrepo4-development-data
```

**Comment by mjgiarlo** [comment]
I believe the previously merged PR updating `solr_wrapper` fixes this for us. I'm not sure caching in Travis is going to pay off in the long run.

**Comment by darrenleeweber** [comment]
Looks like the verdict is to drop the travis-cache for these downloads.  I want to leave this open for a few days, maybe a week, in case there are any dissenting views.  After that, feel free to close it out.  Another PR is going to capture some details (adding .solr_wrapper and .fcrepo_wrapper to hyrax).

**Comment by mjgiarlo** [comment]
Sounds good, @darrenleeweber.

**Comment by atz** [comment]
Right now Travis is timing out of my branch's CI (after a very tiny commit), so **anything** that reduces CI build time should get renewed attention.

**Comment by jcoyne** [comment]
@atz agreed, however this hasn't been shown to do that.

**Pull request closed without merging**